Mathlib/Analysis/Meromorphic/Divisor/Basic.lean

```diff
@@ -0,0 +1,420 @@
+/-
+Copyright (c) 2025 Stefan Kebekus. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stefan Kebekus
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.Analysis.SpecialFunctions.Log.Basic
+
+/-!
+# Divisors on subsets of normed fields
+
+This file defines divisors, a standard book-keeping tool in complex analysis
+used to keep track of pole/vanishing orders of meromorphic objects, typically
+functions or differential forms. It provides supporting API and establishes
+divisors as an instance of a lattice ordered commutative group.
+
+Throughout the present file, `𝕜` denotes a nontrivially normed field, and `U` a
+subset of `𝕜`.
+
+## TODOs
+
+- Constructions: The divisor of a meromorphic function, behavior under product
+  of meromorphic functions, behavior under addition, behavior under restriction
+- Construction: The divisor of a rational polynomial
+-/
+
+open Filter Metric Real Set Topology
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜] {U : Set 𝕜}
+
+/-!
+## Definition, coercion to functions and basic extensionality lemmas
+
+A divisor on `U` is a function `𝕜 → ℤ` whose support is discrete within `U` and
+entirely contained within `U`.  The theorem
+`supportDiscreteWithin_iff_locallyFiniteWithin` shows that this is equivalent to
+the textbook definition, which requires the support of `f` to be locally finite
+within `U`.
+-/
+
+/-- A divisor on `U` is a triple specified below. -/
+structure DivisorOn (U : Set 𝕜) where
+  /-- A function `𝕜 → ℤ` -/
+  toFun : 𝕜 → ℤ
+  /-- A proof that the support of `toFun` is contained in `U` -/
+  supportWithinDomain' : toFun.support ⊆ U
+  /-- A proof the the support is discrete within `U` -/
+  supportDiscreteWithinDomain' : toFun =ᶠ[codiscreteWithin U] 0
+
+/-- A divisor is a divisor on `⊤ : Set 𝕜`. -/
+def Divisor (𝕜 : Type*) [NontriviallyNormedField 𝕜] := DivisorOn (⊤ : Set 𝕜)
```

Comment:
Can you test that `Add` instances, etc., work for `Divisor`? If not, then this need to be `abbrev Divisor ...`.

Reply:
I believe this works, even with the latest changes by Commelin.